Lib/tarfile.py

```diff
@@ -1437,9 +1437,15 @@ def _proc_pax(self, tarfile):
         # these fields are UTF-8 encoded but since POSIX.1-2008 tar
         # implementations are allowed to store them as raw binary strings if
         # the translation to UTF-8 fails.
-        match = re.search(br"\d+ hdrcharset=([^\n]+)\n", buf)
-        if match is not None:
-            pax_headers["hdrcharset"] = match.group(1).decode("utf-8")
+        if (
+            # Statement is both a contains check (!=-1) and a bounds check (>0)
+            (hdrcharset_offset := buf.find(b" hdrcharset=") - 1) > -1
+            # Check that the character before is a digit (0x30-0x39 is 0-9)
+            and 0x30 <= buf[hdrcharset_offset] <= 0x39
+        ):
+            match = re.match(br"^\d{1,20} hdrcharset=([^\n]+)\n", buf[hdrcharset_offset:])
+            if match is not None:
+                pax_headers["hdrcharset"] = match.group(1).decode("utf-8")
```

Comment:
You can siply use `br"\d hdrcharset=(.+?)\n"`.

Reply:
@gpshead I've implemented your suggestion (and added more framing tests) here: https://github.com/python/cpython/pull/121286/commits/39a419cbc6e345da202546af3af4d1be08aeb0a7 Please take a look and let me know what you think.